cloudbridge/cloud/providers/azure/resources.py

```diff
@@ -955,6 +955,91 @@ def create_subnet(self, label, cidr_block, zone=None):
     def gateways(self):
         return self._gateway_service
 
+    def _replace_tag(self, old_tag_name, new_tag_name, new_tag_value):
+        tags = self._network.tags
+        tags.pop(old_tag_name)
+        if new_tag_name and new_tag_value:
+            tags.update({new_tag_name: new_tag_value})
+        self._network.tags = tags
+        self._provider.azure_client. \
+            update_network_tags(self.id, self._network)
+
+    def _existing_sn_tag(self, sn_name):
+        for tag in self._network.tags:
+            if "SubnetLabels" in tag and sn_name in tag:
+                return tag
+        return None
+
+    def _available_sn_tag(self, sn_name, sn_label):
+        max_name = 512 - len(sn_name)
+        max_label = 256 - len(sn_label)
+        for tag in self._network.tags:
+            if "SubnetLabels" in tag:
+                if len(tag) < max_name:
+                    value = self._network.tags.get(tag)
+                    if len(value) < max_label:
+                        return tag
+        return None
+
+    def _get_sn_label(self, sn_name):
+        tag = self._existing_sn_tag(sn_name)
+        if tag:
+            curr_names = tag.split('_')
+            curr_values = tag.split('_')
+            for i in range(curr_names):
+                if curr_names[i] == sn_name:
+                    return curr_values[i]
+        else:
+            return None
+
+    def _remove_sn_label(self, sn_name):
+        tag = self._existing_sn_tag(sn_name)
+        if tag:
+            curr_names = tag.split('_')
+            curr_values = tag.split('_')
+            for i in range(curr_names):
+                if curr_names[i] == sn_name:
+                    curr_names.pop(i)
+                    curr_values.pop(i)
+            if len(curr_names) <= 1:
+                curr_names = []
+                curr_values = []
+            self._replace_tag(tag, "_".join(curr_names), "_".join(curr_values))
+
+    def _replace_sn_label(self, tag, sn_name, new_label):
+        curr_names = tag.split('_')
+        curr_values = tag.split('_')
+        for i in range(curr_names):
+            if curr_names[i] == sn_name:
+                curr_values[i] = new_label
+        self._replace_tag(tag, tag, "_".join(curr_values))
+
+    def _add_sn_label(self, sn_name, sn_label):
+        # Refresh net to make sure tags are updated
+        self.refresh()
+        self.wait_till_ready()
+        # If the subnet is already tagged, replace the current value in the
+        # same tag
+        curr_tag = self._existing_sn_tag(sn_name)
+        if curr_tag:
+            self._replace_sn_label(curr_tag, sn_name, sn_label)
```

Comment:
if two different operations start updating subnet labels on the same network together, we have a potential race condition and potential lost update issue when these values are concatenated. I think the safer option is to revert to having a unique tag per subnet, although that restricts the number of subnets per network to 14 minus any used tags. Alternatively, we just drop label support altogether for subnets. I'm leaning more heavily in favour of the latter.